### PR TITLE
p1: improve bpm bench

### PR DESCRIPTION
**Pull request by skyzh**
DO NOT PUSH PROJECT SOLUTIONS PUBLICLY.

And especially do NOT open pull requests with project solutions!

Please be considerate of this free educational resource.

### CODE REVIEW

File: tools/bpm_bench/bpm_bench.cpp [hunk 1]
```diff
@@ -124,102 +164,133 @@ auto main(int argc, char **argv) -> int {
...
       BpmMetrics metrics(fmt::format("get  {:>2}", thread_id), duration_ms);
       metrics.Begin();
 
       while (!metrics.ShouldFinish()) {
         auto page_idx = dist(gen);
-        auto *page = bpm->FetchPage(page_ids[page_idx], AccessType::Get);
+        auto *page = bpm->FetchPage(page_ids[page_idx], AccessType::Lookup);
         if (page == nullptr) {
           continue;
         }
```
Comment (c8ef) [inline tools/bpm_bench/bpm_bench.cpp:282]:
```suggestion
```suggestion
```suggestion

```
```
```

File: tools/bpm_bench/bpm_bench.cpp [hunk 2]
```diff
@@ -100,6 +96,45 @@ struct BpmMetrics {
...
+auto ModifyPage(char *data, size_t page_idx, uint64_t seed) -> void {
+  auto pg = reinterpret_cast<BustubBenchPageHeader *>(data);
+  pg->seed_ = seed;
+  pg->page_id_ = page_idx;
+  pg->data_[pg->seed_ % 4000] = pg->seed_ % 256;
+}
+
+/// Check the page and verify the data inside
+auto CheckPageConsistentNoSeed(const char *data, size_t page_idx) -> void {
+  auto pg = reinterpret_cast<const BustubBenchPageHeader *>(data);
```
Comment (xzhseh) [inline tools/bpm_bench/bpm_bench.cpp:115]: Seems `auto` will drop the `const` qualifier?
```suggestion
  const auto *pg = reinterpret_cast<const BustubBenchPageHeader *>(data);
```

File: tools/bpm_bench/bpm_bench.cpp [hunk 3]
```diff
@@ -100,6 +96,45 @@ struct BpmMetrics {
...
+  auto right = static_cast<unsigned int>(pg->seed_ % 256);
+  if (left != right) {
+    fmt::println(stderr, "page content not consistent: data_[{}]={} seed_ % 256={}", pg->seed_ % 4000, left, right);
+    std::terminate();
+  }
+}
+
+/// Check the page and verify the data inside
+auto CheckPageConsistent(const char *data, size_t page_idx, uint64_t seed) -> void {
+  auto pg = reinterpret_cast<const BustubBenchPageHeader *>(data);
```
Comment (xzhseh) [inline tools/bpm_bench/bpm_bench.cpp:130]: ditto

**Comment by skyzh** [comment]
TL;DR:

* add more verifications to the buffer pool manager bench so that students cannot workaround it
* better latency policy for disk manager (1ms for random access, 0.1ms for sequential scan)

**Review by fernandolis10 — APPROVED** [review]
LGTM!
